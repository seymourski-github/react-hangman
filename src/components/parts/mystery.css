
.mystery-letter {
  color: #353535;
  flex: 0 0 1em;
  font-size: 1.6em;
  height: 1.23em;
  padding: 0;
  position: relative;
  text-align: center;
  text-indent: -0.1em;
}

.mystery-letter::after {
  content: "";
  border-bottom: 1px #353535 solid;
  width: 90%;
  display: block;
  position: absolute;
  height: inherit;
  top: 0;
}

.mystery-spacer {
  flex: 0 0 0.5em;
}
