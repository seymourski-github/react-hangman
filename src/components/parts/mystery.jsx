import React from 'react';
import PropTypes from 'prop-types';
import { Container, Label } from '../common';
import './mystery.css';

const makeConfig = (letters, activeLetters) => (
  letters.reduce((acc, letter) => {
  acc.push({
    revealed: !activeLetters || activeLetters[letter.toLowerCase()],
    space: letter===' ',
    text: letter
  });
  return acc;
}, []));

export const Mystery = props => {
  const { id, mysteryWord, activeLetters, getClassName } = props;

  // make letter array from mystery word with any revealed/spaces
  const letters = makeConfig(mysteryWord.split(''), activeLetters);

  // only show label text if already revealed
  const labels = letters.map((item, i) => (
    <Label {...item}
      key={`${item.text}${i}`}
      className={getClassName(item.text, props)}
      children={item.revealed&&!item.space ? item.text: ''}
      text={item.revealed&&!item.space ? item.text: ''}
    />));

  return <Container flex={true} id={id} children={labels} />;

};

Mystery.propTypes = {
  activeLetters: PropTypes.object,
  mysteryWord: PropTypes.string,
  id: PropTypes.string.isRequired,
  getClassName: PropTypes.func
};

Mystery.defaultProps = {
  activeLetters: {},
  mysteryWord: '',
  getClassName: (str, props) => {
    const space = str===' ';
    return `label ${space?'mystery-spacer':'mystery-letter'}`;
  }
};
