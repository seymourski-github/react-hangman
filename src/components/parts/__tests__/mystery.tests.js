import React from 'react';
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import { Mystery } from '../mystery';

const props = {
  id: 'my-mystery-panel',
  activeWord: 'My secret word',
  activeLetters: {
    e: true
  }
};

describe('<Mystery /> Component', function() {

  const component = renderer.create(<Mystery {...props} />);
  let wrapper = shallow(<Mystery {...props} />);

  it("renders and matches our snapshot", () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
