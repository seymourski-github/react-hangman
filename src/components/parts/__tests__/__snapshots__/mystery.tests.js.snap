// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Mystery /> Component renders and matches our snapshot 1`] = `
<div
  className="container row"
  id="my-mystery-panel"
  style={undefined}
>
  <label
    className="label mystery-letter"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    
  </label>
  <label
    className="label mystery-letter"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    
  </label>
  <label
    className="label mystery-spacer"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    
  </label>
  <label
    className="label mystery-letter"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    
  </label>
  <label
    className="label mystery-letter"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    e
  </label>
  <label
    className="label mystery-letter"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    
  </label>
  <label
    className="label mystery-letter"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    
  </label>
  <label
    className="label mystery-letter"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    e
  </label>
  <label
    className="label mystery-letter"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    
  </label>
  <label
    className="label mystery-spacer"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    
  </label>
  <label
    className="label mystery-letter"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    
  </label>
  <label
    className="label mystery-letter"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    
  </label>
  <label
    className="label mystery-letter"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    
  </label>
  <label
    className="label mystery-letter"
    disabled={false}
    hidden={false}
    htmlFor={undefined}
    style={
      Object {
        "alignSelf": "",
        "flex": undefined,
        "visibility": "visible",
      }
    }
  >
    
  </label>
</div>
`;
